### PR TITLE
feat(native): add param-free pre-norm (LNPre/RMSPre) and fold_ln support

### LINKED ISSUE #1570: [Proposal] Native adapter: param-free pre-norm (LNPre/RMSPre) + fold_ln support
### Proposal
Base on `dev-4.x`. Teach the native bridge adapter to build param-free pre-norms and to fold LayerNorm, in `transformer_lens/model_bridge/sources/native/model.py` and `transformer_lens/model_bridge/supported_architectures/native.py`.

### Motivation
Deprecating `HookedTransformer` requires its toy-model demos (`demos/Othello_GPT.ipynb`, in the CI notebook matrix) to run on `TransformerBridge`, which needs `normalization_type="LNPre"` + folding.

### Pitch
- When `normalization_type` is `"LNPre"`/`"RMSPre"`, build a param-free norm (no learnable weight/bias) instead of `nn.LayerNorm`: update `_make_norm` (`model.py:69-74`), `_make_norm_bridge` (`native.py:39-44`), and `_init_norm` (`native/init.py:98-107`).
- Today `"RMSPre"` silently aliases to `"RMS"` (`sources/native/model.py:41`, `supported_architectures/native.py:28`) and `"LNPre"` falls through to plain `nn.LayerNorm`. Both silently produce a *parameterized* norm rather than erroring, so the fix must also stop the silent aliasing.
- Enable folding on the native adapter: populate `weight_processing_conversions` and flip `supports_fold_ln` / `supports_center_writing_weights` (`native.py:98-100`).

Pointers:
- `_init_norm` (`sources/native/init.py:98-107`) raises `TypeError` on unknown norm classes, so the new param-free class must be registered there or init hard-fails.
- Evidence for the fold_ln half: `supported_architectures/native.py:94-100` carries an explicit rationale comment — "Compatibility-mode fold_ln / center_writing_weights aren't wired up... gate the corresponding ProcessWeights paths off".

Acceptance:
- [ ] `boot_native` LNPre builds a param-free norm matching HT `LayerNormPre`
- [ ] Folded native forward matches unfolded within fp tolerance
- [ ] New tests under `tests/unit/model_bridge/`
- [ ] `make unit-test` passes
- [ ] `uv run mypy .` clean

Mirror: `HookedTransformer.process_weights_` / `fold_layer_norm` in `transformer_lens/HookedTransformer.py`; `LayerNormPre` in `transformer_lens/components/`.

### Checklist
- [x] I have checked that there is no similar issue in the repo (**required**)

### PR DESCRIPTION
**Pull request by MdSadiqMd**
# Description

Add param-free pre-norm support (LNPre/RMSPre) and enable fold_ln on the native bridge adapter.

**Problem:** Silent aliasing where "RMSPre" was treated as "RMS" (creating parameterized norms) and "LNPre" fell through to `nn.LayerNorm`. Both silently produced parameterized norms instead of param-free ones.

**Solution:**
- Add `NativeRMSNormPre` and `NativeLayerNormPre` param-free norm classes
- Update `_make_norm` to distinguish RMS vs RMSPRE, LN vs LNPRE
- Enable `supports_fold_ln` and `supports_center_writing_weights` on native adapter

**Motivation:** Deprecating HookedTransformer requires its toy-model demos (`demos/Othello_GPT.ipynb`, in the CI notebook matrix) to run on TransformerBridge, which needs `normalization_type="LNPre"` + folding.

Fixes #1570 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### Screenshots
N/A

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

File: transformer_lens/model_bridge/supported_architectures/native.py [hunk 1]
```diff
@@ -92,11 +102,10 @@ def __init__(self, cfg: Any) -> None:
...
 
         # Native layout already stores Q/K/V split; no rearranges needed.
-        # Compatibility-mode fold_ln / center_writing_weights aren't wired up,
-        # so gate the corresponding ProcessWeights paths off — folding without
-        # the state-dict conversions would mis-place or drop weights.
-        self.supports_fold_ln = False
-        self.supports_center_writing_weights = False
+        # Support fold_ln (LN weight folds into downstream layers) and
+        # center_writing_weights (residual-stream-writing weights are centered).
+        self.supports_fold_ln = True
```
Comment (jlarson4) [inline transformer_lens/model_bridge/supported_architectures/native.py:112]: Setting `supports_fold_ln` true turns on a path that mis-folds: with non-identity LN weights the folded model's logits diverge relative to the unfolded model, and `enable_compatibility_mode()` on an `LNPre` model crashes with an `EinopsError` inside `fold_layer_norm`. The root cause is that `weight_processing_conversions` stays `{}`, so `convert_tensor_to_tl_format` passes native `(out, in)` Linear weights through unchanged while the fold formulas assume TL `[head, d_model, d_head]` format. Issue #1570 called for populating the conversions alongside the flag flip. Could you wire the conversions? The base adapter ships a shared helper `_qkvo_weight_conversions` that other split-QKV adapters use.

File: transformer_lens/model_bridge/supported_architectures/native.py [hunk 2]
```diff
@@ -37,7 +43,11 @@ def _is_rotary(cfg: Any) -> bool:
 
 
 def _make_norm_bridge(name: str, cfg: Any, *, force_rms: bool = False):
-    if force_rms or _uses_rms(cfg):
+    norm_type = (getattr(cfg, "normalization_type", None) or "LN").upper()
+
+    if _uses_param_free_norm(cfg):
```
Comment (jlarson4) [inline transformer_lens/model_bridge/supported_architectures/native.py]: Returning a plain `GeneralizedComponent` for LNPre/RMSPre drops `hook_scale` and `hook_normalized`. HookedTransformer's `LayerNormPre` exposes those attributes and `ActivationCache.apply_ln_to_stack` consumes them. `run_with_cache` on an LNPre native model records zero scale/normalized hooks and `ActivationCache.apply_ln_to_stack` raises `KeyError` on `ln_final.hook_scale`. Could the pre-norms get a norm-shaped bridge component that keeps these hooks, the way `NormalizationBridge` does?

File: transformer_lens/model_bridge/sources/native/model.py [hunk 3]
```diff
@@ -66,11 +66,47 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+
+
+class NativeLayerNormPre(nn.Module):
+    """Param-free LayerNorm — center + normalize only, no learnable scale/bias."""
+
+    def __init__(self, eps: float = 1e-5):
+        super().__init__()
+        self.eps = eps
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (jlarson4) [inline transformer_lens/model_bridge/sources/native/model.py:94]: `NativeRMSNormPre` normalizes in fp32 and casts back (lines 76–80), matching HT `LayerNormPre`'s reduced-precision handling, but `NativeLayerNormPre` computes its mean and variance in the input dtype. Should the two new classes share the same dtype policy so half-precision native models normalize consistently?

File: transformer_lens/model_bridge/sources/native/model.py [hunk 4]
```diff
@@ -66,11 +66,47 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        x = x - x.mean(dim=-1, keepdim=True)
+        scale = (x.pow(2).mean(dim=-1, keepdim=True) + self.eps).sqrt()
+        return x / scale
+
+
 def _make_norm(cfg: TransformerBridgeConfig, *, force_rms: bool = False) -> nn.Module:
-    if force_rms or _uses_rms_norm(cfg):
+    norm_type = _normalization_type(cfg)
+
+    if force_rms or norm_type in ("RMS", "RMSPRE"):
```
Comment (jlarson4) [inline transformer_lens/model_bridge/sources/native/model.py:111]: With `normalization_type="LNPre"` and `final_rms=True`, this branch builds a parameterized `NativeRMSNorm` for `ln_out`, but HookedTransformer's contract for that config is the param-free `RMSNormPre`. The bridge side already treats it as param-free (`supported_architectures/native.py:48` fires before its `force_rms` check), so the two sides now disagree about whether `ln_out` has weights. Could this branch return the param-free RMS norm whenever the config is a pre-norm type, so the module matches both the bridge wrapper and HT?

File: tests/unit/model_bridge/test_boot_native.py [hunk 5]
```diff
@@ -323,3 +430,29 @@ def test_boot_native_supports_training_step():
     ), "No non-zero gradients after backward"
     optimizer.step()
     optimizer.zero_grad()
+
+
+def test_boot_native_fold_ln_output_invariant():
```
Comment (jlarson4) [inline tests/unit/model_bridge/test_boot_native.py:435]: This test cannot fail. `boot_native` initializes norms to identity, making the scale-fold a no-op, and the 0.01 loss tolerance hides the logit error the current fold introduces. Can we randomize the norm parameters and compare at the logit level, similar to how `tests/unit/test_weight_processing.py:838-877` works? It would also be nice to add a LNPre-config case, since that workflow currently crashes.

**Review by jlarson4 — COMMENTED** [review]
Hey @MdSadiqMd, thanks for taking this on!

The param-free norm half of this PR is solid. The math is good, and the code is well structured. The `fold_ln` half is not quite there yet, and I have a handful of requests below. Let me know if you have any questions!

**Comment by MdSadiqMd** [comment]
Hey @jlarson4 implemented the changes. Let me know if anything is required from my side

**Comment by jlarson4** [comment]
Looks great! Merging now, thank you @MdSadiqMd!
